app/models/classification.rb

```diff
@@ -579,6 +583,13 @@ def delete_tags_and_entries
     delete_tag_and_taggings
   end
 
+  def validate_tag_mapping
+    if tag && ProviderTagMapping.exists?(:tag_id => tag.id)
```

Comment:
is there a Rails association between a tag and a ProviderTagMapping?  If so, I'd expect something like

```ruby
if tag&.provider_tag_mappings.exist?
```

Reply:
We don't have that association but it can be added. I think it needs to be a `has_many` because mapping multiple different labels to the same category is supported. This change seems to work -
```diff
diff --git a/app/models/classification.rb b/app/models/classification.rb
index f6b7c242e6..d49ce7d9e7 100644
--- a/app/models/classification.rb
+++ b/app/models/classification.rb
@@ -584,7 +584,7 @@ class Classification < ApplicationRecord
   end

   def validate_tag_mapping
-    if tag && ProviderTagMapping.exists?(:tag_id => tag.id)
+    if tag&.provider_tag_mappings.present?
       errors.add("", _("A Tag Mapping exists for this category and must be removed before deleting"))
       throw :abort
     end
diff --git a/app/models/provider_tag_mapping.rb b/app/models/provider_tag_mapping.rb
index d0ef7a3c5a..e6fe64580d 100644
--- a/app/models/provider_tag_mapping.rb
+++ b/app/models/provider_tag_mapping.rb
@@ -19,6 +19,7 @@
 #
 class ProviderTagMapping < ApplicationRecord
   belongs_to :tag
+  has_many   :provider_tag_mappings

   require_nested :Mapper
```
I can push that change if you think it's worth it

Reply:
`.present?` does a full query and then a `.present?` call on the resultant array, which is inefficient.  `.any?` or `.exists?` does `select 1 from` which is a deeper presence query.

That being say, I didn't realize the provider_tag_mappings would be on ProviderTagMapping...is it nested?  I'm just confused how that works?

Reply:
@Fryguy  Sorry, the change to `provider_tag_mapping.rb` was a mistake. I meant to put it on `tag` and that class already has the 1has_many` relationship. So just the change with `any?` would suffice, as you suggested
```diff
diff --git a/app/models/classification.rb b/app/models/classification.rb
index f6b7c242e6..63fae82f1d 100644
--- a/app/models/classification.rb
+++ b/app/models/classification.rb
@@ -584,7 +584,7 @@ class Classification < ApplicationRecord
   end

   def validate_tag_mapping
-    if tag && ProviderTagMapping.exists?(:tag_id => tag.id)
+    if tag&.provider_tag_mappings.any?
       errors.add("", _("A Tag Mapping exists for this category and must be removed before deleting"))
       throw :abort
     end
```